tests/host_tools/fcmetrics.py

```diff
@@ -437,53 +437,40 @@ def get_emf_unit_for_fc_metrics(full_key):
-    Use an existing metrics logger with existing dimensions so
-    that its easier to corelate the metrics with the test calling it.
-    Add a prefix "fc_metrics." to differentiate these metrics, this
-    also helps to avoid using this metrics in AB tests.
+    Flush Firecracker metrics to CloudWatch. Use an existing metrics logger with existing dimensions so that it is
+    easier to correlate the metrics with the test calling it. Add a prefix "fc_metrics." to differentiate these metrics,
+    this also helps to avoid using this metrics in A/B tests.
     NOTE:
-        There are metrics with keywords "fail", "err",
-        "num_faults", "panic" in their name and represent
-        some kind of failure in Firecracker.
-        This function `does not` assert on these failure metrics
-        since some tests might not want to assert on them while
-        some tests might want to assert on some but not others.
+        There are metrics with keywords "fail", "err", "num_faults", "panic" in their name and represent some kind of
+        failure in Firecracker.  We assert that all these are zero, to catch potentially silent failure modes. This
+        means the FcMonitor cannot be used in negative tests that might cause such metrics to be emitted.
     """
 
-    def walk_key(full_key, keys):
-        for key, value in keys.items():
-            final_full_key = full_key + "." + key
-            if isinstance(value, dict):
-                walk_key(final_full_key, value)
-            else:
-                # values are 0 when:
-                # - there is no update
-                # - device is not used
-                # - SharedIncMetric reset to 0 on flush so if
-                #   there is no change metric the values remain 0.
-                # We can save the amount of bytes we export to
-                # CloudWatch in these cases.
-                # however it is difficult to differentiate if a 0
-                # should be skipped or upload because it could be
-                # an expected value in some cases so we upload
-                # all the metrics even if data is 0.
-                unit = get_emf_unit_for_fc_metrics(final_full_key)
-                metrics.put_metric(f"fc_metrics.{final_full_key}", value, unit=unit)
-
-    # List of SharedStoreMetric that once updated have the same value thoughout the life of vm
-    metrics_to_export_once = {
-        "api_server",
-        "latencies_us",
-    }
-    skip = set()
-    for group, keys in fc_metrics.items():
-        if group == "utc_timestamp_ms":
+    # Pre-order tree traversal to convert a tree into its list of paths with dot separate segments
+    def flatten_dict(node):
+        if not isinstance(node, dict):
+            return {"": node}
+
+        return {
+            "." + child_key + full_key: leaf
+            for child_key, child_node in node.items()
+            for full_key, leaf in flatten_dict(child_node).items()
+        }
```

Comment:
I think this is more readable:
```python
ret = {}
for child_key, child_node in node.items():
  for full_key, leaf in flatten_dict(child_node).items():
    key = "." + child_key + full_key
    ret[key] = leaf
```
Also not sure if this is a `pre-order` as we never visit the root nodes, but I guess this does not really matter.

Reply:
Less pythonic, but applied your suggestion xP

btw, yeah, the way I had it originally written it was pre order (the full paths were constructed top down, with prefixes passed down to the recursive calls). Now it's actually post-order (we first do the recursive call, and then construct the path strings bottom up as suffixes). Not sure why I changed this though, actually, the prefix thingy results in less weird string concatenations, I think

Reply:
Changed it back to preorder, its indeed more readable this way. Wonder what I missed last friday